File: example/js/iui/keyboard.js
'use strict';

// Portable IUI keyboard readiness check and one consistent key snapshot.
//
// Hardware-free by design: a machine without a readable keyboard reports that through the
// readiness state instead of failing. Build with WSE_BUILD_IUI=ON and WSE_BUILD_NODE_BINDING=ON.
//
// The argument is either the installed `lang/js` package directory or a built `wse.node` addon.
// Given the addon the sample still loads the package, so what is exercised is the public
// JavaScript surface an application uses rather than the raw native entry points.
//
// Two things the binding does not carry: the C++ counterpart reports through
// wse::registDefaultLog() and wse::WLog(), which have no JavaScript equivalent, so this sample
// writes to the console; and a Keyboard offers no event or callback surface at all, so readiness
// is polled and the key state is sampled.
//
//   node example/js/iui/keyboard.js <installed lang/js directory or wse.node>
//
// This sample takes no arguments of its own.

const path = require('node:path');

const target = process.argv[2] || process.env.WSE_JAVASCRIPT_PACKAGE;
if ( !target ) {
  throw new Error('Pass the installed lang/js directory, or the built wse.node addon.');
}
// A `.node` argument is the build-tree addon. Point the package loader at it and still require the
// package, so the sample always demonstrates the same public surface an application uses.
if ( target.endsWith('.node') ) process.env.WSE_NODE_ADDON = path.resolve(target);
const wse = require(target.endsWith('.node')
  ? path.resolve(__dirname, '..', '..', '..', 'lang', 'js')
  : path.resolve(target));

if ( wse.Keyboard === undefined ) {
  throw new Error('This WSE build does not include the IUI component.');
}

// Every WSE enumeration reaches JavaScript as a plain object of number-valued properties, not as a
// type, so a readable report needs the reverse map built here.
const stateNames = Object.fromEntries(
  Object.entries(wse.KeyboardAccessState).map(([name, value]) => [value, name]),
);

// A group arrives as a fixed-length array of booleans, one entry per key; wse.KeyboardGroupSize
// carries the length of each group for a caller that wants to index it by key rather than count.
function countPressed(group) {
  return group.reduce((total, pressed) => total + (pressed ? 1 : 0), 0);
}

async function main() {
  // Construction starts the backend's monitoring, and this object is its only owner. JavaScript
  // has no deterministic destructor, so the try/finally below is what releases it: skip close()
  // and the monitor keeps running until the garbage collector finalizes the object, at an
  // unpredictable moment or, in a short-lived script, not at all before exit.
  const keyboard = new wse.Keyboard();
  try {
    // The backend needs a moment before it reports its first reading.
    // Twenty attempts of 25 ms is a half-second ceiling, generous for every backend measured. A
    // shorter budget would report Starting on a slow machine and skip the snapshot; a longer one
    // only delays the same answer where the keyboard is genuinely unavailable. wse.wait() is the
    // binding's own timer: it resolves a promise, so awaiting it yields the loop rather than
    // blocking it.
    for ( let attempt = 0; attempt < 20 && !keyboard.isAvailable(); ++attempt ) {
      await wse.wait(25);
    }

    const state = keyboard.accessState();
    console.log('keyboard state:', stateNames[state] ?? state);

    if ( !keyboard.isAvailable() ) {
      // Not a failure of this sample: snapshot() would throw a WseError explaining why.
      // Unavailable, PermissionDenied, and Disconnected all arrive here, and each read re-checks
      // the readiness state before touching the device, so nothing partial is ever returned.
      // Returning normally is deliberate; a machine with no readable keyboard is a supported
      // environment for this sample, and the process still exits with success.
      try {
        keyboard.snapshot();
      } catch (error) {
        console.log('keyboard is not readable:', error.message);
      }
      return;
    }

    // One snapshot is one consistent point in time; do not combine several separate reads.
    const snapshot = keyboard.snapshot();
    console.log('pressed ascii keys:', countPressed(snapshot.ascii));
    console.log('pressed function keys:', countPressed(snapshot.function));
    console.log('pressed arrow keys:', countPressed(snapshot.arrow));
    console.log('pressed lock keys:', countPressed(snapshot.lock));
    console.log('pressed command keys:', countPressed(snapshot.command));
    console.log('pressed ascii code:', keyboard.pressedAscii());
  } finally {
    // close() releases the native monitor and is safe to call twice, but it is terminal for a
    // Keyboard: unlike WebCamera.close(), which only ends a session, every later call on this
    // object throws "The keyboard is already closed.". Create a new Keyboard to read again.
    keyboard.close();
  }
}

// A failure raised by WSE carries the structured triple printed here. A plain TypeError from the
// binding's argument checks does not, and shows as undefined in the first three fields.
main().catch((error) => {
  console.error(error.category, error.code, error.nativeCode, error.message);
  process.exitCode = 1;
});

File: example/python/iui/keyboard.py
"""Portable IUI keyboard readiness check and one consistent key snapshot.

Hardware-free by design: a machine without a readable keyboard reports that through the
readiness state instead of failing. Build with WSE_BUILD_IUI=ON and WSE_BUILD_PYTHON_BINDING=ON.

Constructing a Keyboard starts a monitoring thread, so readiness is not immediate: a caller must
poll `is_available` or `access_state` before its first read rather than assume the first snapshot
will work. `close()`, which `__exit__` calls, drops the native keyboard and ends that thread at a
point the caller chose; without it the thread would live until the interpreter collected the
object. There is no reopen: a closed Keyboard raises on every later call, so a second reading
needs a second object.

The C++ counterpart reports through the WSE log facility. The binding exposes no logging, so this
sample prints; it also has no way to turn a readiness state into a structured error on its own, so
it lets `snapshot()` raise and reads the reason off the exception.

With no argument the sample imports the installed `wse` package; with an argument it loads the
built extension module from that path instead.

    python example/python/iui/keyboard.py [path to the built _wse module]
"""

import importlib.util
import sys
import time
from pathlib import Path


def load_wse():
    if len(sys.argv) == 1:
        import wse

        return wse
    module_path = Path(sys.argv[1]).resolve()
    spec = importlib.util.spec_from_file_location("_wse", module_path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Unable to load WSE Python module: {module_path}")
    module = importlib.util.module_from_spec(spec)
    sys.modules[spec.name] = module
    spec.loader.exec_module(module)
    return module


def main() -> int:
    wse = load_wse()
    # A build without WSE_BUILD_IUI registers no Keyboard at all, so the name has to be looked for
    # rather than caught; there is no stub that raises a friendlier error later.
    if not hasattr(wse, "Keyboard"):
        raise SystemExit("This WSE build does not include the IUI component.")

    # The context manager is the single owner; leaving it stops the monitoring thread.
    with wse.Keyboard() as keyboard:
        # The backend needs a moment before it reports its first reading. Twenty tries at 25 ms is
        # a half-second ceiling: long enough that a working keyboard is never reported as STARTING,
        # short enough that a machine which will never have one is not made to wait. The loop exits
        # on the first success, so the ceiling only costs time when there is nothing to read.
        for _ in range(20):
            if keyboard.is_available:
                break
            time.sleep(0.025)

        print("keyboard state:", keyboard.access_state)

        if not keyboard.is_available:
            # Not a failure of this sample: snapshot() explains why the keyboard is unreadable.
            # UNAVAILABLE, PERMISSION_DENIED, and DISCONNECTED all land here, and each is a real
            # state of the machine rather than a defect. Returning 0 keeps a headless or sandboxed
            # runner green; branch on the error's category and code if a caller needs to tell them
            # apart, never on the message text.
            try:
                keyboard.snapshot()
            except wse.WseError as failure:
                print("keyboard is not readable:", failure)
            return 0

        # One snapshot is one consistent point in time; do not combine several separate reads.
        # It is a plain dictionary of bool lists owned by the caller, with the fixed lengths
        # KEYBOARD_ASCII_COUNT, KEYBOARD_FUNCTION_COUNT and friends, so it stays readable after the
        # keyboard is closed. pressed_ascii() below is a separate read and can disagree with it.
        snapshot = keyboard.snapshot()
        for group in ("ascii", "function", "arrow", "lock", "command"):
            print(f"pressed {group} keys:", sum(snapshot[group]))
        print("pressed ascii code:", keyboard.pressed_ascii())

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
